### ISSUE TITLE
question-encoder output error

### ISSUE THREAD

**Issue by 2020zyc** [facebookresearch/DPR#85]
### Background
- I use the ***train_dense_encoder.py*** to train a biencoder with my own data. Training data are constructed as the readme says.

- Configuration is as follows:
![image](https://user-images.githubusercontent.com/7539692/100985376-e04d8a80-3586-11eb-8aa8-be88ade253dd.png)

- After the loss goes down to about 3.3, it can't go down;  If I don't use hard-negative-samples, it gets ~2.3.

### Error 
- I save the biencoder.question_model use ***generate_dense_embeddings.py***. 
- Then load it using BertModel.from_pretrained(xxx), BertModel is a kind of Huggingface model. Also the tokenizer is the BertTokenizer from Huggingface.
- I test the loaded BertModel using four completely different samples. The last hidden state in each position of each sample are the same. All states of [CLS] have no difference.
![image](https://user-images.githubusercontent.com/7539692/100987038-ee040f80-3588-11eb-8b07-2e91ba137d83.png)
- So the retrived docs also the same between all four samples.

Any help, thanks.

**Comment by 2020zyc** [comment]
@vlad-karpukhin Can you hlep me? Thanks

**Comment by vlad-karpukhin** [comment]
2x12 as the batch size is pretty low for a high quality model.
Static output from a trained encoder seems really strange and looks like your trining set the model to some trivial state.
What was your correct prediction ratio for NLL loss validation?

**Comment by 2020zyc** [comment]
> 2x12 as the batch size is pretty low for a high quality model.
> Static output from a trained encoder seems really strange and looks like your trining set the model to some trivial state.
> What was your correct prediction ratio for NLL loss validation?

I now use 3 gpus for distributed training, each 24samples/batch. With 30/40 epochs, the training loss currently is 1.3 and can't decrease any more. 

And 
NLL Validation: **"NLL Validation: loss = 1.714351. correct prediction ratio  5323/9984 ~  0.533153"**
Average Rank Validation: ***3gpus, val_av_rank_other_neg=20, val_av_rank_hard_neg=1, dev_batch_size=16***
```bash
Average rank validation ...
Average rank validation ...
Average rank validation ...
Av.rank validation: step 99, computed ctx_vectors 35200, q_vectors 1600
Av.rank validation: step 99, computed ctx_vectors 35200, q_vectors 1600
Av.rank validation: step 99, computed ctx_vectors 35200, q_vectors 1600
Av.rank validation: step 199, computed ctx_vectors 70400, q_vectors 3200
Av.rank validation: step 199, computed ctx_vectors 70400, q_vectors 3200
Av.rank validation: step 199, computed ctx_vectors 70400, q_vectors 3200
Av.rank validation: total q_vectors size=torch.Size([3328, 384])
Av.rank validation: total ctx_vectors size=torch.Size([73216, 384])
Av.rank validation: total q_vectors size=torch.Size([3328, 384])
Av.rank validation: total ctx_vectors size=torch.Size([73216, 384])
Av.rank validation: total q_vectors size=torch.Size([3328, 384])
Av.rank validation: total ctx_vectors size=torch.Size([73216, 384])
Av.rank validation: average rank 2411.587439903846, total questions=9984
Av.rank validation: average rank 2411.587439903846, total questions=9984
Av.rank validation: average rank 2411.587439903846, total questions=9984
```

**Comment by vlad-karpukhin** [comment]
Yes, it is quite bad, 
one should expect correct prediction ratio to be ~ 90% and it should pretty quickly (like 1 epoch) jump to > 70% values. But that is for 8x16=128 batch size when correct predcition is to select the right ctx out of 128 (x2=256 actually due to 1 hard negative per question) candidates. In your case the number of candidates 3x16x2=96 which a way easier task than to select from 256. So the training should give even higher numbers for correct prediction ratio in your case.
Looks like some training issue, but no idea what may be wrong. You can try to play with learning rate though; and also with gradient accumulation steps parameter but there is a known bug in our code with that (https://github.com/facebookresearch/DPR/issues/31).

This doesn't explain degrading to some trivial state though but might be related.

Even if you manage to fix that bug and and find a better hyper-parameters, I'd still expect the model quality to be inferior to ours and maybe even to BM25 since effective batch size is a critical parameter.
